### PR TITLE
TCES-51 tim timeline backend api

### PR DESCRIPTION
**Pull request by kevinle623**
# Description

Implemented the timeline backend api for employers, job leads, and clients

Issue # 51

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist

- [x] My code follows the style guidelines of this project (required)
- [x] I have performed a self-review of my code (required)
- [x] I have linted and formatted my code (required)
- [x] New and existing unit tests pass locally with my changes (required)

### CODE REVIEW

File: server/src/models/employer_timeline_entry.model.js
Comment (DanielDervishi) [inline server/src/models/employer_timeline_entry.model.js]: I don't think we need entries associated with these two types, job_lead_add should be sufficient

File: server/src/controllers/job_lead_timeline_entries/addJobLeadTimelineEntry.js
Comment (DanielDervishi) [inline server/src/controllers/job_lead_timeline_entries/addJobLeadTimelineEntry.js]: Correct me if I'm wrong but I think you need the await keyword in front of this method since its asyncronous

File: server/src/controllers/job_lead/addJobLeads.js
Comment (DanielDervishi) [inline server/src/controllers/job_lead/addJobLeads.js]: Employer should be specified here too

File: server/src/controllers/client/updateClient.js
Comment (DanielDervishi) [inline server/src/controllers/client/updateClient.js]: I don't think this query is correct (should be able to access the user through req.user)

File: server/src/controllers/client/updateClient.js
Comment (DanielDervishi) [inline server/src/controllers/client/updateClient.js]: Both the first name and last name should be displayed

File: server/src/controllers/client/updateClient.js
Comment (DanielDervishi) [inline server/src/controllers/client/updateClient.js:105]: Its probably best for status_at_exit, status_at_x_months updates to be displayed as well

File: server/src/controllers/job_lead/addJobLeads.js
Comment (DanielDervishi) [inline server/src/controllers/job_lead/addJobLeads.js:33]: We should be creating job lead timeline entries even on bulk create

File: server/src/controllers/job_lead_timeline_entries/addJobLeadTimelineEntry.js
Comment (DanielDervishi) [inline server/src/controllers/job_lead_timeline_entries/addJobLeadTimelineEntry.js]: Also the employer should be passed in to submitPlacementUpdateEntryIntTimelines

**Review by DanielDervishi — COMMENTED** [review]
Hey Kevin, great work! Just left a few comments and I'll take another look at this once those are addressed

File: server/src/controllers/job_lead_timeline_entries/addJobLeadTimelineEntry.js
Comment (kevinle623) [reply server/src/controllers/job_lead_timeline_entries/addJobLeadTimelineEntry.js]: good catch

File: server/src/models/client_timeline_entry.model.js
Comment (VangoCode) [inline server/src/models/client_timeline_entry.model.js:21]: On PlanetScale, we defined this enum as also possibly being `add`. Should it also be here, or was there a change in the db schema?

File: server/src/models/employer_timeline_entry.model.js
Comment (VangoCode) [inline server/src/models/employer_timeline_entry.model.js:23]: Similar to my previous comment, this enum doesn't include everything defined on PlanetScale -- is that schema out-of-date?

File: server/src/models/employer_timeline_entry.model.js
Comment (VangoCode) [inline server/src/models/employer_timeline_entry.model.js]: I believe this should not be allowed to be null (otherwise, we need to update the table schema)
(retrospective note, I noticed that there is a defaultValue being set, so it will never be null but I think we should still set this to false to be clearer)

File: server/src/models/employer_timeline_entry.model.js
Comment (VangoCode) [inline server/src/models/employer_timeline_entry.model.js:63]: Similar to above, I believe this should not be allowed to be null

File: server/src/models/employer_timeline_entry.model.js
Comment (VangoCode) [inline server/src/models/employer_timeline_entry.model.js:77]: `allowNull` is not specified here, so it defaults to `true`, but I believe the expected behaviour is that a user should be specified

File: server/src/models/job_lead_timeline_entry.model.js
Comment (VangoCode) [inline server/src/models/job_lead_timeline_entry.model.js:50]: We will need to update the PlanetScale schema to include this, as it currently does not have an `employer` column

**Review by VangoCode — COMMENTED** [review]
Hi Kevin! Overall this looks great, awesome work 😄 . I left a few clarifying comments related to the database. Otherwise, it all looks good to me, and I am good to approve after they've been discussed!

File: server/src/models/client_timeline_entry.model.js
Comment (kevinle623) [reply server/src/models/client_timeline_entry.model.js:21]: Thanks Ron, I made changes to the model and forgot to update the schema on planet scale good catch!

File: server/src/controllers/client/updateClient.js
Comment (jordanjanakievski) [inline server/src/controllers/client/updateClient.js]: Let's replace `req.body.values.user` with `req.user` anywhere it is located since the user info should automatically be attached to the request

File: server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js
Comment (jordanjanakievski) [inline server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js:7]: For pagination we probably also want it to be time based. 

Consider this abstracted list of timeline entries [update 4, update 3, update 2, update 1] -- page 1 is currently [update 4 and update 3] for page size 2. Imagine another update happens creating a new timeline entry. Now the list would be [update 5, update 4, update 3, update 2, update 1]. When a tces employee clicks next page now they will see [update 3, update 2], update being on both pages.

Since many actions can happen in a short period this is a concern. Also note that this applies to each getAllObjectTimelineEntries controller method.

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (jordanjanakievski) [inline server/src/controllers/client_timeline_entries/addClientTimelineEntry.js:6]: When creating a new timeline entry, we should be generating the title and body from info passed in through the request rather than relying on the frontend to generate the title and body entirely

**Review by jordanjanakievski — COMMENTED** [review]
Great work Kevin! Just left a few comments relating to how responsibilities when creating a timeline entry should be shared as well as a few other things. I'll take another look once these are addressed 
- Daniel from Jordan's computer

File: server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js
Comment (kevinle623) [reply server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js:7]: Hi, @DanielDervishi  just to confirm 
```
Consider this abstracted list of timeline entries [update 4, update 3, update 2, update 1] -- page 1 is currently [update 4 and update 3] for page size 2. Imagine another update happens creating a new timeline entry. Now the list would be [update 5, update 4, update 3, update 2, update 1]. When a tces employee clicks next page now they will see [update 3, update 2], update being on both pages.
```

 would be the expected behaviour right? If so, this should be the current behaviour. Was wondering if I can focus on this in the timeline routing ticket where I will interact more with the pagination from the frontend side and then will make adjustments to backend in that ticket.

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (kevinle623) [reply server/src/controllers/client_timeline_entries/addClientTimelineEntry.js:6]: Hi @DanielDervishi, I was planning on doing the auto generation on the timeline routing ticket, just so that when I work on connecting the frontend I can picture all the possible entries I need to autogenerate from in the frontend, it will be hard to do it now with purely / take some time to consider all the needed cases now to auto generate. Was hoping to focus this in the timeline routing ticket to make sure I cover all the possible timeline entry creation. I hope this is fine.

Comment (DanielDervishi) [reply server/src/controllers/client_timeline_entries/addClientTimelineEntry.js:6]: Hey Kevin, that works as well

File: server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js
Comment (DanielDervishi) [reply server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js:7]: Hey Kevin, I guess my explanation was a bit bad since the timeline doesn't exactly follow a page model the same way a table may. The point is if we have something like infinite scrolling to populate the timeline, we don't want new entries being added causing other entries showing up twice. Using the previous example, the timeline (on the frontend) at the beginning may have entries [update 4, update 3]. Then a new entry is added called "update 5". After the user scrolls down in order to see more timeline entires, [update 4, update 3, update 2, update 1] should be visible in the timeline. If we use a standard pagination model without timestamps, accessing those additional results using page 2 of the getAllObjectTimelineEntries request at this point would be [update 3, update 2] since update 5 was added. This would cause the timeline to have entries [update 4, update 3, update 3, update 2] if you just append the results from page 2 of getAllObjectTimelineEntries to the existing list of timeline entries on the frontend.

Of course though I'll leave it up to you to decide how you want to achieve the intended behaviour

Comment (kevinle623) [reply server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js:7]: Oh I see makes sense @DanielDervishi , yeah I think this will be dependent on the how the frontend is implemented. Because for the timeline entries table, to solve the issue that you talked about above I was not thinking of doing paging, I was thinking we can just fetch the size of the time table so far, so if the user keeps scrolling down, then we increment the amount of entries we want further. For example lets say we are initially at 10 entries showing, we scroll down, then we request 20 entries, keep scrolling down we keep 30 entries, this eensures that every time the user scrolls down, we get the most up to date most recent  x entries. Let me know what you think of this approach. But nonetheless, was wondering if this can be focused on the timeline routing ticket as well?

Comment (DanielDervishi) [reply server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js:7]: Ok that makes sense, the approach definitely works I just don't know how smooth scrolling will be with it since we are essentially building the whole timeline from scratch whenever a user scrolls down as we are fetching existing entries as well

Comment (kevinle623) [reply server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js:7]: @DanielDervishi I was thinking when the user scrolls to the bottom, we prompt a "Load More" button within the timeline to load 10 more.

Comment (DanielDervishi) [reply server/src/controllers/client_timeline_entries/getAllClientTimelineEntries.js:7]: Ok sounds good to me

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (DanielDervishi) [inline server/src/controllers/client_timeline_entries/addClientTimelineEntry.js]: We probably shouldn't allow an update entry to be created here since they are created automatically in updateClient

File: server/src/controllers/employer_timeline_entries/addEmployerTimelineEntry.js
Comment (DanielDervishi) [inline server/src/controllers/employer_timeline_entries/addEmployerTimelineEntry.js]: The job_lead_add entry shouldn't be necessary here since it is automatically added upon job lead creation.

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (DanielDervishi) [inline server/src/controllers/client_timeline_entries/addClientTimelineEntry.js]: Employer should be extracted from the job lead and be passed in here

File: server/src/utils/placement_entry_util.js
Comment (DanielDervishi) [inline server/src/utils/placement_entry_util.js:15]: user is now a user object, so this query won't work

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (DanielDervishi) [inline server/src/controllers/client_timeline_entries/addClientTimelineEntry.js]: Is there a reason we are expecting a title and body on placement? I think these are being generated automatically by submitPlacementUpdateEntryInTimelines. This is also the case in the other timelines I believe

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (DanielDervishi) [inline server/src/controllers/client_timeline_entries/addClientTimelineEntry.js]: This should be the user id rather than the user object

File: server/src/controllers/client/updateClient.js
Comment (DanielDervishi) [inline server/src/controllers/client/updateClient.js]: client should be client.id here

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (DanielDervishi) [reply server/src/controllers/client_timeline_entries/addClientTimelineEntry.js]: Also placement should probably be removed from the client timeline api since you can't place a client from their timeline in the designs. Rather the entry in the client timeline should be created when the client is placed through the employer or job lead timeline.

File: server/src/models/employer_timeline_entry.model.js
Comment (DanielDervishi) [inline server/src/models/employer_timeline_entry.model.js]: We probably shouldn't throw an error here since it is pretty common to not specify a client (for example when creating a note). The same is the case for job_lead or in any other timeline model where one of the associated objects isn't always passed in.

File: server/src/controllers/employer_timeline_entries/addEmployerTimelineEntry.js
Comment (DanielDervishi) [inline server/src/controllers/employer_timeline_entries/addEmployerTimelineEntry.js]: user -> `user: user.id`

**Review by DanielDervishi — COMMENTED** [review]
Hey Kevin, sorry for the wait. After these comments are addressed we should be good to merge the PR in 😄

File: server/src/utils/placement_entry_util.js
Comment (kevinle623) [reply server/src/utils/placement_entry_util.js:15]: im passing in user as the id not the object into this function.

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (kevinle623) [reply server/src/controllers/client_timeline_entries/addClientTimelineEntry.js]: user variable is the user id

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (kevinle623) [reply server/src/controllers/client_timeline_entries/addClientTimelineEntry.js:37]: I wasn't trying to make it so robust especially since this is only the backend api implementation, I was planning on polishing it and removing what I don't need in the frontend implementation to save a bit of time now so I can get started with the frontend ticket faster

File: server/src/controllers/employer_timeline_entries/addEmployerTimelineEntry.js
Comment (kevinle623) [reply server/src/controllers/employer_timeline_entries/addEmployerTimelineEntry.js]: user is the user id variable fetched from req.user

**Comment by kevinle623** [comment]
> Hey Kevin, sorry for the wait. After these comments are addressed we should be good to merge the PR in 😄

@DanielDervishi Hello Daniel, no worries thanks for the review. I have address the comments and made the needed changes when you have time to take a look again. In general, I was hoping to focus more on the robustness of the backend and polishing the actions we really do need when I start with  the timeline routing ticket.

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (DanielDervishi) [reply server/src/controllers/client_timeline_entries/addClientTimelineEntry.js]: Screenshot 2024-03-03 at 4 10 51 PM
Since we access user from the request it is a user object here

File: server/src/controllers/employer_timeline_entries/addEmployerTimelineEntry.js
Comment (DanielDervishi) [reply server/src/controllers/employer_timeline_entries/addEmployerTimelineEntry.js]: req.user is actually the user object itself, not the id

File: server/src/utils/placement_entry_util.js
Comment (DanielDervishi) [reply server/src/utils/placement_entry_util.js:15]: Ok that sounds good, but currently as its passed in within addEmployerTimelineEntry and addJobLeadTimelineEntry it is the user object that is being passed in not the user id (following from req.user returning the user object and not id of the user)
Screenshot 2024-03-03 at 4 14 50 PM

File: server/src/controllers/client_timeline_entries/addClientTimelineEntry.js
Comment (kevinle623) [reply server/src/controllers/client_timeline_entries/addClientTimelineEntry.js]: sounds good i have swtiched it out.

**Comment by kevinle623** [comment]
@DanielDervishi, thanks for the clarification about the user id, I have switched it out so that I treat the initial const {user} = req as the User object itself and have pushed the changes.

**Review by DanielDervishi — APPROVED** [review]
LGTM!

**Review by jordanjanakievski — APPROVED** [review]
Looks great! Thank you Kevin!

**Pull request merged**
